beluga_tutorial/src/main.cpp

```diff
@@ -0,0 +1,258 @@
+          static_cast<double>(std::get<1>(resample)));
+    }
+
+    node["simulation_dataset"][cycle]["estimation"]["mean"] = std::get<0>(data.estimation);
+    node["simulation_dataset"][cycle]["estimation"]["sd"] = std::get<1>(data.estimation);
+  }
+
+  const std::string bags_path{"./src/beluga/beluga_tutorial/bags/"};
+  std::filesystem::create_directory(bags_path);
+  std::ofstream fout(bags_path + file_name, std::ios::trunc);
+  fout << node;
+  fout.close();
+}
+
+int main() {
+  // Load params from yaml file
+  TutorialParams tutorial_params;
+  load_params_from_yaml(tutorial_params);
+
+  // Create the TutorialDataset object for record the simulation data
+  TutorialDataset tutorial_dataset;
+
+  // Load map from file or generate a random map
+  LandmarkMapVector landmark_map;
+  if (tutorial_params.load_landmark_map_from_file) {
+    load_landmark_map_from_yaml(landmark_map);
+  } else {
+    std::uniform_int_distribution landmark_distribution{0, tutorial_params.map_size};
+    landmark_map = beluga::views::sample(landmark_distribution) |
+                   ranges::views::take_exactly(tutorial_params.number_of_doors) | ranges::to<LandmarkMapVector>;
+    landmark_map |= ranges::actions::sort | ranges::actions::unique;
+  }
+
+  tutorial_dataset.landmark_map = landmark_map;
+
+  // Generate particles
+  std::normal_distribution<double> initial_distribution(tutorial_params.initial_pose, tutorial_params.initial_pose_sd);
+  auto particles = beluga::views::sample(initial_distribution) |
+                   ranges::views::transform(beluga::make_from_state<Particle>) |
+                   ranges::views::take_exactly(tutorial_params.number_of_particles) | ranges::to<beluga::TupleVector>;
+
+  // Execute the particle filter using beluga
+  double current_pose{tutorial_params.initial_pose};
+  for (auto n = 0; n < tutorial_params.number_of_cycles; n++) {
+    // Prepare the data to be saved
+    TutorialData tutorial_data;
+
+    // Save ground truth
+    current_pose += (tutorial_params.velocity * tutorial_params.dt);
+    tutorial_data.ground_truth = current_pose;
+
+    // Check if the simulation is out of bounds
+    if (current_pose > tutorial_params.map_size) {
+      break;
+    }
+
+    // Motion model
+    auto motion_model = [&](double state) {
+      double distance = (tutorial_params.velocity * tutorial_params.dt);
+      double translation_param = generateRandom(0.0, tutorial_params.translation_sigma);
+      return state + distance - translation_param;
+    };
+
+    // Generate simulated sensor data
+    auto sensor_data = landmark_map | ranges::views::remove_if([&](const double lm) {
+                         return std::abs(lm - current_pose) > tutorial_params.sensor_range;
+                       }) |
+                       ranges::views::transform([&](const double lm) { return lm - current_pose; }) |
+                       ranges::to<SensorData>;
```

Comment:
@Alondruck meta: we are feeding perfect range measurements, aren't we? There's no noise here.

Reply:
Yes, you're right. In the sensor data, there's no noise. However, we can say that we have noise when we do the correspondence between the data from the sensor and the data of the landmarks relative to each particle. Actually, they are different types of noises, so I can add noise to the sensor data if it's necessary.